File: go/internal/copilot/shell_executor.go
package copilot

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strings"
)

type ShellExecutor struct {
	reader *bufio.Reader
}

func NewShellExecutor() *ShellExecutor {
	return &ShellExecutor{
		reader: bufio.NewReader(os.Stdin),
	}
}

func (s *ShellExecutor) ExecuteWithConfirmation(command, explanation string) (bool, error) {
	fmt.Printf("\nSuggestion: %s\n", command)
	fmt.Printf("Explanation: %s\n", explanation)
	fmt.Print("\nExecute this command? [y/N] ")

	input, err := s.reader.ReadString('\n')
	if err != nil {
		return false, err
	}

	answer := strings.ToLower(strings.TrimSpace(input))
	if answer == "y" || answer == "yes" {
		fmt.Printf("Executing: %s...\n", command)

		var cmd *exec.Cmd
		if runtime.GOOS == "windows" {
			cmd = exec.Command("cmd", "/C", command)
		} else {
			cmd = exec.Command("sh", "-c", command)
		}

		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr

		err := cmd.Run()
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		}
		return true, nil
	}

	fmt.Println("Command execution cancelled.")
	return false, nil
}

File: java/src/main/java/com/tormentnexus/copilot/ShellExecutor.java
package com.tormentnexus.copilot;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Scanner;

public class ShellExecutor {
    private final Scanner scanner;

    public ShellExecutor() {
        this.scanner = new Scanner(System.in);
    }

    public boolean executeWithConfirmation(String command, String explanation) {
        System.out.println("\nSuggestion: " + command);
        System.out.println("Explanation: " + explanation);
        System.out.print("\nExecute this command? [y/N] ");

        String answer = scanner.nextLine().trim().toLowerCase();

        if (answer.equals("y") || answer.equals("yes")) {
            System.out.println("Executing: " + command + "...");
            try {
                ProcessBuilder processBuilder = new ProcessBuilder();
                if (System.getProperty("os.name").toLowerCase().contains("win")) {
                    processBuilder.command("cmd.exe", "/c", command);
                } else {
                    processBuilder.command("sh", "-c", command);
                }

                Process process = processBuilder.start();

                BufferedReader stdInput = new BufferedReader(new InputStreamReader(process.getInputStream()));
                BufferedReader stdError = new BufferedReader(new InputStreamReader(process.getErrorStream()));

                String s;
                while ((s = stdInput.readLine()) != null) {
                    System.out.println(s);
                }
                while ((s = stdError.readLine()) != null) {
                    System.err.println(s);
                }

                process.waitFor();
                return true;
            } catch (Exception e) {
                System.err.println("Error executing command: " + e.getMessage());
                return false;
            }
        }

        System.out.println("Command execution cancelled.");
        return false;
    }
}